diff --git a/shell-command/src/shell_detect.rs b/shell-command/src/shell_detect.rs
index 0fb73df38..96d8d338b 100644
--- a/shell-command/src/shell_detect.rs
+++ b/shell-command/src/shell_detect.rs
@@ -60,6 +60,12 @@ pub fn detect_shell_type(shell_path: impl AsRef<std::path::Path>) -> Option<Shel
 
 #[cfg(unix)]
 fn get_user_shell_path() -> Option<PathBuf> {
+    if let Some(shell_path) =
+        std::env::var_os("SHELL").filter(|shell_path| !shell_path.is_empty())
+    {
+        return Some(shell_path.into());
+    }
+
     let uid = unsafe { libc::getuid() };
     use std::ffi::CStr;
     use std::mem::MaybeUninit;
